File: Python/Samples/Proxy/DelayPricerProxy.py
import sys
import math
from decimal import *
import os.path
sys.path.append(os.path.join(os.path.dirname(__file__), '../Decorator/'))
from StdPricers import Pricer
from Models import Transport
from ServicePricers import DelayPricer


class DelayPricerProxy(Pricer):
    """延遲計費
    """

    MAX_DELAY_HOURS = 2

    def __init__(self,  pricer=Pricer):
        super().__init__(pricer.customer, pricer.receiver, pricer.freight)
        self.delayPricer = DelayPricer(pricer)

    def price(self, transport=Transport):
        """Return Total Price"""
        totalPrice = 0
        servicePrice = 0
        exceedMaxDelayHours = 0

        if(transport.delayHours <= self.MAX_DELAY_HOURS):
            totalPrice = self.delayPricer.price(transport)
        else:
            exceedMaxDelayHours = transport.delayHours - self.MAX_DELAY_HOURS  # 計算超過的小時
            transport.delayHours = self.MAX_DELAY_HOURS
            totalPrice = self.delayPricer.price(transport)

        servicePrice = exceedMaxDelayHours * 1000
        totalPrice += servicePrice
        print("延遲(超過兩小時)服務費用 = {0}，總費用={1}".format(servicePrice, totalPrice))
        return totalPrice

File: CSharp/DP.Domain/Samples/Proxy/DelayPricerProxy.cs
using System.Diagnostics;
using DP.Domain.Samples.Decorator;

namespace DP.Domain.Samples.Proxy
{
    public class DelayPricerProxy : IPricer
    {
        public string Customer { get; set; }
        public string Receiver { get; set; }
        public string Freight { get; set; }
        private IPricer _delayPricer = null;
        private readonly int MAX_DELAY_HOURS = 2;

        public DelayPricerProxy(IPricer pricer)
        {
            this._delayPricer = new DelayPricer(pricer);
        }
        public decimal Price(Transport transport)
        {
            decimal totalPrice = 0;
            decimal servicePrice = 0;
            var exceedMaxDelayHours = 0;
            
            if(transport.DelayHours<=MAX_DELAY_HOURS) //未超過
            {
                totalPrice = this._delayPricer.Price(transport);
            }
            else
            {
                exceedMaxDelayHours = transport.DelayHours - MAX_DELAY_HOURS;//計算超過的小時
                transport.DelayHours = MAX_DELAY_HOURS;
                totalPrice = this._delayPricer.Price(transport);
            }
            
            servicePrice = exceedMaxDelayHours * 1000;
            totalPrice += servicePrice;
            Trace.WriteLine($"延遲(超過兩小時)服務費用 = {servicePrice}，總費用={totalPrice}");
            return totalPrice;
        }
    }
}